### ISSUE TITLE
Handle SIGPIPE for Network Target

### ISSUE THREAD

**Issue by matthew-macgregor** [goatshriek/stumpless#231]
### TL;DR

The network target will terminate the program on POSIX systems if the remote closes the connection. I think the simplest solution is to pass the `MSG_NOSIGNAL` flag to the `send()` function. This should prevent the signal and is scope to just that call.

Might be a very minor change (a few lines?), but this proposed solution needs to be validated.

Other approaches are listed below.

### Details

**Is your feature request related to a problem? Please describe.**

On POSIX systems, a broken pipe or closed socket will result in the SIGPIPE signal being raised, and the default behavior of the signal is to terminate the process. The stumpless TCP network target (among others?) doesn't handle SIGPIPE, which will kill the program if raised.

- Stumpless version: v2.0.0
- Platform: Any POSIX compatible OS? Observed on Ubuntu Linux 22.04 (5.15.0-47-generic).

I put together an MRE: https://github.com/matthew-macgregor/stumpless-network-sigpipe-example

**Workaround**

In the short term, the programmer should be aware of this behavior when using the network target, and can either `signal( SIGPIPE, SIG_IGN )` to ignore the signal, or use the more advanced signal handling functions to temporarily block then retrieve signals if more control is needed. Using a global signal handler is discouraged, in part because we can't be sure that stumpless is in a consistent state from inside the handler (and because there are better solutions).

**Describe the solution you'd like**

Since the underlying function `sys_socket_sendto_target` is using the `send()` function, `MSG_NOSIGNAL` should be passed in the flags argument.

This should prevent the `SIGPIPE` for just that send and return -1  (which is the expected behavior) and set errno to `EPIPE`.

This should be portable on POSIX systems, but `winsock2.h` doesn't support this flag, so some OS-specific code may be necessary.

- https://www.freebsd.org/cgi/man.cgi?query=send
- https://linux.die.net/man/2/send
- https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-send

**Caveats**

- My proposal is based on some research but I'm no expert on signal handling and this proposal should be evaluated by someone with more experience!
- Windows doesn't have signals, so this code would only apply to POSIX systems.

**Describe alternatives you've considered**

Globally ignoring the signal isn't appropriate for a library, since the programmer may need control of this signal. The same is true for a global signal handler, since it's not easy to reason about reentrancy.

Stumpless could delegate signal handling to the integrator (and provide documentation), but this feels like a detail that the library target should handle. Bottom line, a logging framework should never cause the process to be killed unexpectedly.

Another option:

Stumpless could treat the send to a network target as a critical section, block the signal while sending *on the current thread*, get pending signals, and restore the original signal mask.

1. Use `pthread_sigmask` to add `SIGPIPE` to the mask just prior to sending data. Store the original mask.
2. Send the data.
3. Use `sigtimedwait` to get pending `SIGPIPE` signals.
4. Restore the original mask.

This should work, but seems a lot more complicated than the solution I proposed. It could be a fallback approach if it turns out that I'm wrong about `MSG_NOSIGNAL`.

See the following for more detail: 

- https://linux.die.net/man/3/pthread_sigmask
- http://www.microhowto.info/howto/ignore_sigpipe_without_affecting_other_threads_in_a_process.html

**Additional context**

- Discussion: https://github.com/goatshriek/stumpless/issues/78#issuecomment-1229197064
- Runnable MRE: https://github.com/matthew-macgregor/stumpless-network-sigpipe-example
